README.md
# react-leaflet-marker-layer

`react-leaflet-marker-layer` provides a simple `<MarkerLayer />` component for plotting React components as markers in a `react-leaflet` map.

![A screenshot of markers on a leaflet map](../master/screenshot.jpg?raw=true)

## Usage

```js
import React from 'react';
import { render } from 'react-dom';
import { Map, Marker, Popup, TileLayer } from 'react-leaflet';
import MarkerLayer from '../src/MarkerLayer';

const position = { lng: -122.673447, lat: 45.522558 };
const markers = [
  {
    position: { lng: -122.67344700000, lat: 45.522558100000 },
    text: 'Voodoo Doughnut',
  },
  {
    position: { lng: -122.67814460000, lat: 45.5225512000000 },
    text: 'Bailey\'s Taproom',
  },
  {
    position: { lng: -122.67535700000002, lat: 45.5192743000000 },
    text: 'Barista'
  },
  {
    position: { lng: -122.65596570000001, lat: 45.5199148000001 },
    text: 'Base Camp Brewing'
  }
];

class ExampleMarkerComponent extends React.Component {

  render() {
    const style = {
      border: 'solid 1px lightblue',
      backgroundColor: '#333333',
      borderRadius: '50%',
      marginTop: '-12px',
      marginLeft: '-12px',
      width: '24px',
      height: '24px'
    };

    return (
      <div style={Object.assign({}, this.props.style, style)}></div>
    );
  }

}

class MapExample extends React.Component {

  state = {
    mapHidden: false
  };

  render() {
    if (this.state.mapHidden) {
      return (
        <div>
          <input type="button" value="Toggle Map" onClick={() => this.setState({ mapHidden: !this.state.mapHidden })} />
        </div>
      );
    }

    return (
      <div>
        <Map center={position} zoom={13}>
          <MarkerLayer
            markers={markers}
            longitudeExtractor={m => m.position.lng}
            latitudeExtractor={m => m.position.lat}
            markerComponent={ExampleMarkerComponent} />
          <TileLayer
            url='http://{s}.tile.osm.org/{z}/{x}/{y}.png'
            attribution='&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors'
          />
        </Map>
        <input type="button" value="Toggle Map" onClick={() => this.setState({ mapHidden: !this.state.mapHidden })} />
      </div>
    );
  }

}

render(<MapExample />, document.getElementById('app'));
```

## API

The `MarkerLayer` component takes the following props:

- `markers`: an array of objects to be plotted
- `longitudeExtractor`: a function that returns the marker object's longitude e.g. `marker => marker.lng`
- `latitudeExtractor`: a function that returns the marker object's latitude e.g. `marker => marker.lat`
- `markerComponent`: (required) the React component to be rendered for each marker, this component will receive the following props
  - `marker`: the object from the `markers` array
  - `style`: a style object for positioning, you should include these styles on your component
  - `map`: the Leaflet map object from the `react-leaflet` `MapLayer`
  - `...propsForMarkers`: the component will also receive the properties of `propsForMarkers` as props
- `propsForMarkers`: props to pass on to marker components

## Example

To try the example:

1. Clone this repository
2. run `npm install` in the root of your cloned repository
3. run `npm run example`
4. Visit [localhost:8000](http://localhost:8000)

## Contributing

See [CONTRIBUTING.md](https://www.github.com/OpenGov/react-leaflet-marker-layer/blob/master/CONTRIBUTING.md)

## License

`react-leaflet-marker-layer` is MIT licensed.

See [LICENSE.md](https://www.github.com/OpenGov/react-leaflet-marker-layer/blob/master/LICENSE.md) for details.


CHANGELOG.md
# 0.1.3 Release

- Only fit bounds on update when data or position extractors change

# 0.1.2 Release

- Improve performance by implementing `shouldComponentUpdate` with `react-pure-render/function`

# 0.1.1 Release

- fix casing issue with `isNumber` import

# 0.1.0 Release

- Handle fitBounds better when bad data is present
- added `fitBoundsOnUpdate` parameter to fit the markers on the map when they change

# 0.0.6 Release

- Actually build the 0.0.5 changes

# 0.0.5 Release

- Marker objects that return bad data for location are no longer rendered and are ignored when repositioning

# 0.0.2 Release

- Fix an issue with unmounting layer

# 0.0.1 Release

- Initial implementation of package


src/__tests__/MarkerLayer.test.js
import React from 'react';
import { shallow, mount, render } from 'enzyme';
import { Map } from 'react-leaflet';
import MarkerLayer from '../MarkerLayer.js';

jest.unmock('../MarkerLayer.js');

const L = jest.genMockFromModule('leaflet');

class MarkerComponent extends React.Component {
  render() {
    return (
      <div className="marker-component">Marker component</div>
    );
  }
}

describe('MarkerLayer', () => {

  /*
    Here we declare mocks or fixtures of
    all the data structures that the component uses
  */
  const center = { lng: -122.673447, lat: 45.522558 };

  const mockBounds = {
    contains: jest.fn(() => true),
    extend: jest.fn(),
    getNorthEast: jest.fn(() => ({ lng: -122, lat: 46 })),
    getSouthWest: jest.fn(() => ({ lng: -123, lat: 45 })),
  };

  const mockPanes = { overlayPane: { appendChild: jest.fn() } };

  const mockMap = {
    layerPointToLatLng: jest.fn(() => ({ lng: -122.6, lat: 45.522 })),
    latLngToLayerPoint: jest.fn(() => ({ x: 100, y: 100 })),
    on:  jest.fn(),
    getBounds:  jest.fn(() => mockBounds),
    getPanes: jest.fn(() => mockPanes),
    invalidateSize: jest.fn()
  };

  const mockMarkers = [
    {
      position: { lng: -122.673447, lat: 45.5225581 },
      text: 'Voodoo Doughnut',
    },
    {
      position: { lng: -122.6781446, lat: 45.5225512 },
      text: 'Bailey\'s Taproom',
    },
    {
      position: { lng: -122.67535700000002, lat: 45.5192743 },
      text: 'Barista'
    }
  ];

  it('should render', () => {
    const layer = render(
        <MarkerLayer
          markers={[]}
          longitudeExtractor={m => m.position.lng}
          latitudeExtractor={m => m.position.lat}
          markerComponent={MarkerComponent} />
    );
    expect(layer).toBeTruthy();
  });

  it('should render a single child <MarkerComponent /> given one marker', () => {
    const layer = mount(
        <MarkerLayer
          map={mockMap}
          markers={[mockMarkers[0]]}
          longitudeExtractor={m => m.position.lng}
          latitudeExtractor={m => m.position.lat}
          markerComponent={MarkerComponent} />
    );
    expect(layer.find('.marker-component').length).toEqual(1);
  });

  it('should render three child <MarkerComponent /> given three markers with the different positions', () => {
    const layer = mount(
        <MarkerLayer
          map={mockMap}
          markers={mockMarkers}
          longitudeExtractor={m => m.position.lng}
          latitudeExtractor={m => m.position.lat}
          markerComponent={MarkerComponent} />
    );
    expect(layer.find('.marker-component').length).toEqual(3);
  });

  it('should pass the `propsForMarkers` prop to rendered <MarkerComponent />', () => {
    const mockProps = {
      theAnswer: 42,
      numCoffees: 3
    };

    const layer = mount(
        <MarkerLayer
          map={mockMap}
          markers={mockMarkers}
          longitudeExtractor={m => m.position.lng}
          latitudeExtractor={m => m.position.lat}
          propsForMarkers={mockProps}
          markerComponent={MarkerComponent} />
    );
    const componentProps = layer.find(MarkerComponent).at(0).props();
    expect(componentProps).toBeTruthy();
    expect(componentProps.theAnswer).toEqual(mockProps.theAnswer);
    expect(componentProps.numCoffees).toEqual(mockProps.numCoffees);
  });
});


src/MarkerLayer.js
import React from 'react';
import ReactDOM from 'react-dom';
import map from 'lodash.map';
import forEach from 'lodash.foreach';
import pluck from 'lodash.pluck';
import min from 'lodash.min';
import max from 'lodash.max';
import isNumber from 'lodash.isnumber';
import filter from 'lodash.filter';
import L from 'leaflet';
import { MapLayer } from 'react-leaflet';
import shouldPureComponentUpdate from 'react-pure-render/function';

export type LngLat = {
  lng: number;
  lat: number;
}

export type Point = {
  x: number;
  y: number;
}

export type Map = {
  layerPointToLatLng: (lngLat: Point) => LngLat;
  latLngToLayerPoint: (lngLat: LngLat) => Point;
  on: (event: string, handler: () => void) => void;
  getPanes: () => Panes;
  invalidateSize: () => void;
}

export type Panes = {
  overlayPane: Pane;
}

export type Pane = {
  appendChild: (element: Object) => void;
}

function isInvalid(num: number): boolean {
  return !isNumber(num) && !num;
}

function isValid(num: number): boolean {
  return !isInvalid(num);
}

function shouldIgnoreLocation(loc: LngLat): boolean {
  return isInvalid(loc.lng) || isInvalid(loc.lat);
}

export default class MarkerLayer extends MapLayer {
  static propTypes = {
    markers: React.PropTypes.array,
    longitudeExtractor: React.PropTypes.func.isRequired,
    latitudeExtractor: React.PropTypes.func.isRequired,
    markerComponent: React.PropTypes.func.isRequired,
    propsForMarkers: React.PropTypes.object,
    fitBoundsOnLoad: React.PropTypes.bool,
    fitBoundsOnUpdate: React.PropTypes.bool,
  };

  shouldComponentUpdate = shouldPureComponentUpdate;

  componentWillReceiveProps() {
    // no-op to override MapLayer behavior
  }

  componentDidMount(): void {
    this.leafletElement = ReactDOM.findDOMNode(this.refs.container);
    this.props.map.getPanes().overlayPane.appendChild(this.leafletElement);
    if (this.props.fitBoundsOnLoad) {
      this.fitBounds();
    }
    this.attachEvents();
    this.updatePosition();
  }

  componentWillUnmount(): void {
    this.props.map.getPanes().overlayPane.removeChild(this.leafletElement);
  }

  fitBounds(): void {
    const markers = this.props.markers;
    const lngs = filter(map(markers, this.props.longitudeExtractor), isValid);
    const lats = filter(map(markers, this.props.latitudeExtractor), isValid);
    const ne = { lng: max(lngs), lat: max(lats) };
    const sw = { lng: min(lngs), lat: min(lats) };

    if (shouldIgnoreLocation(ne) || shouldIgnoreLocation(sw)) {
      return;
    }

    this.props.map.fitBounds(L.latLngBounds(L.latLng(sw), L.latLng(ne)));
  }

  markersOrPositionExtractorsChanged(props): boolean {
    return this.props.markers !== props.markers
      || this.props.longitudeExtractor !== props.longitudeExtractor
      || this.props.latitudeExtractor !== props.latitudeExtractor;
  }

  componentDidUpdate(prevProps): void {
    this.props.map.invalidateSize();
    if (this.props.fitBoundsOnUpdate && this.markersOrPositionExtractorsChanged(prevProps)) {
      this.fitBounds();
    }
    this.updatePosition();
  }

  attachEvents(): void {
    const map: Map = this.props.map;
    map.on('viewreset', () => this.updatePosition());
  }

  getLocationForMarker(marker): LngLat {
    return {
      lat: this.props.latitudeExtractor(marker),
      lng: this.props.longitudeExtractor(marker)
    };
  }

  updatePosition(): void {
    forEach(this.props.markers, (marker, i) => {
      const markerElement = ReactDOM.findDOMNode(
        this.refs[this.getMarkerRefName(i)]
      );

      const location = this.getLocationForMarker(marker);

      if (shouldIgnoreLocation(location)) {
        return;
      }

      const point = this.props.map.latLngToLayerPoint(L.latLng(location));

      L.DomUtil.setPosition(markerElement, point);
    });
  }

  render(): React.Element {
    return (
      <div ref="container"
        className={`leaflet-objects-pane
           leaflet-marker-pane
           leaflet-zoom-hide
           react-leaflet-marker-layer`} >
        {this.renderMarkers()}
      </div>
    );
  }

  renderMarkers(): Array<React.Element> {
    const style = { position: 'absolute' };
    const MarkerComponent = this.props.markerComponent;
    return map(this.props.markers, (marker, index: number) => {
      if (shouldIgnoreLocation(this.getLocationForMarker(marker))) {
        return;
      }

      return (
        <MarkerComponent
          {...this.props.propsForMarkers}
          key={index}
          style={style}
          map={this.props.map}
          ref={this.getMarkerRefName(index)}
          marker={marker} />
      );
    });
  }

  getMarkerRefName(index: number): string {
    return `marker-${index}`;
  }

}


example/index.js
import React from 'react';
import { render } from 'react-dom';
import { Map, Marker, Popup, TileLayer } from 'react-leaflet';
import MarkerLayer from '../src/MarkerLayer';

const position = { lng: -122.673447, lat: 45.522558 };
const markers = [
  {
    position: { lng: -122.67344700000, lat: 45.522558100000 },
    text: 'Voodoo Doughnut',
  },
  {
    position: { lng: -122.67814460000, lat: 45.5225512000000 },
    text: 'Bailey\'s Taproom',
  },
  {
    position: { lng: -122.67535700000002, lat: 45.5192743000000 },
    text: 'Barista'
  },
  {
    position: { lng: -122.65596570000001, lat: 45.5199148000001 },
    text: 'Base Camp Brewing'
  },
  {
    position: { lng: null, lat: 45.522558100000 },
    text: 'A Malformed Location',
  },
];

const markerToBeAdded = {
  position: { lng: -119.0638890000001, lat: 40.883056000001 },
  text: 'Black Rock City'
};

class ExampleMarkerComponent extends React.Component {

  render() {
    const style = {
      border: 'solid 1px lightblue',
      backgroundColor: '#333333',
      borderRadius: '50%',
      marginTop: '-12px',
      marginLeft: '-12px',
      width: '24px',
      height: '24px'
    };

    return (
      <div style={Object.assign({}, this.props.style, style)}></div>
    );
  }

}

const longitudeExtractor = m => m.position.lng;
const latitudeExtractor = m => m.position.lat;

class MapExample extends React.Component {

  state = {
    mapHidden: false,
    layerHidden: false,
    markers: markers,
    msg: null
  };

  _moveMarker() {
    setTimeout(() => {
      markers[0].position.lng = markers[0].position.lng + 2;
      this.setState({ markers: Array.from(markers) });
    }, 1000 * 1);
  }

  componentDidMount() {
    this._moveMarker();
    setTimeout(() => {
      this.setState({ msg: 'LOL' });
    }, 1000 * 5);
  }

  render() {
    if (this.state.mapHidden) {
      return (
        <div>
          <input type="button" value="Toggle Map" onClick={() => this.setState({ mapHidden: !this.state.mapHidden })} />
        </div>
      );
    }

    return (
      <div>
        <Map center={position} zoom={13}>
          {!this.state.layerHidden && <MarkerLayer
            markers={this.state.markers}
            longitudeExtractor={longitudeExtractor}
            latitudeExtractor={latitudeExtractor}
            markerComponent={ExampleMarkerComponent}
            fitBoundsOnUpdate />}
          <TileLayer
            url='http://{s}.tile.osm.org/{z}/{x}/{y}.png'
            attribution='&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors'
          />
        </Map>
        <input type="button" value="Toggle Map" onClick={() => this.setState({ mapHidden: !this.state.mapHidden })} />
        <input type="button" value="Toggle Layer" onClick={() => this.setState({ layerHidden: !this.state.layerHidden })} />
        {!!this.state.msg && this.state.msg}
      </div>
    );
  }

}

render(<MapExample />, document.getElementById('app'));
